src/jit/optimizer.cpp

```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
                 }
 #endif // FEATURE_EH_FUNCLETS && defined(_TARGET_ARM_)
 
-                /* At this point we have a loop - record it in the loop table
+                // Compact the loop (sweep through it and move out any blocks that aren't part of the
+                // flow cycle), and count the exits.
+                BasicBlock* moveAfter = nullptr;
+                for (BasicBlock* previous = top->bbPrev; previous != bottom;)
+                {
+                    BasicBlock* block = previous->bbNext;
```

Comment:
> > I wonder if this would be a clearer if it was conceptually split up...

> I agree the function is more monolithic than ideal...  the amount of state that would need to get passed in and particularly that would need to get passed out,..

Of course, I could wrap that state up in a class and pass it around implicitly.  @AndyAyersMS, let me know if you think https://github.com/JosephTremoulet/coreclr/commit/0861de1cb4efc026cf533759fa8c735938a9adbe is an improvement and I can clean it up and push it to the PR branch.